code/gf128/Hacl.Spec.GF128.PolyLemmas.fst

```diff
@@ -0,0 +1,385 @@
+  assert (r == p % pow2 64);
+  lemma_bitwise_all ();
+  lemma_mask_is_mod a 64;
+  lemma_equal (P.mod a (monomial 64)) (of_uint 128 (p % pow2 64));
+  ()
+#pop-options
+
+#push-options "--z3rlimit 50 --max_fuel 1"
+val lemma_eq_poly_uint_shift_right (a:poly) (p:U.uint_t 128) (s:nat{s <= 64}) : Lemma
+  (requires degree a <= 127 /\ a == of_uint 128 p)
+  (ensures P.shift a (-s) == of_uint 128 (p / pow2 s))
+
+let lemma_eq_poly_uint_shift_right a p s =
+  reveal_defs ();
+  lemma_bitwise_all ();
+  lemma_equal (P.shift a (-s)) (of_uint 128 (U.shift_right p s));
+  ()
+#pop-options
+
+val lemma_eq_poly_uint_shift_left (a:poly) (p:U.uint_t 128) (s:nat{s <= 64}) : Lemma
+  (requires degree a <= 63 /\ a == of_uint 128 p)
+  (ensures P.shift a s == of_uint 128 ((p * pow2 s) % pow2 128))
+
+let lemma_eq_poly_uint_shift_left a p s =
+  lemma_bitwise_all ();
+  lemma_equal (P.shift a s) (of_uint 128 (U.shift_left p s));
+  ()
+
+val lemma_eq_poly_uint_add (a b:poly) (p1 p2:U.uint_t 128) : Lemma
+  (requires degree a <= 127 /\ degree b <= 127 /\
+    a == of_uint 128 p1 /\ b == of_uint 128 p2)
+  (ensures P.add a b == of_uint 128 (U.logxor p1 p2))
+
+let lemma_eq_poly_uint_add a b p1 p2 =
+  lemma_bitwise_all ();
+  lemma_equal (P.add a b) (of_uint 128 (U.logxor p1 p2));
+  ()
+
+val to_vec_mod_pow2: #n:size_nat -> a:U.uint_t n -> m:pos -> i:nat{n - m <= i /\ i < n} ->
+  Lemma (requires (a % pow2 m == 0))
+        (ensures  (index #bool #n (to_vec a) i == false))
+        [SMTPat (index #bool #n (to_vec #n a) i); SMTPat (pow2 m)]
+let rec to_vec_mod_pow2 #n a m i =
+  if i = n - 1 then
+    begin
+    lemma_index_app2 (to_vec #(n - 1) (a / 2)) (Seq.create 1 (a % 2 = 1)) i;
+    mod_mult_exact a 2 (pow2 (m - 1))
+    end
+  else
+    begin
+    lemma_index_app1 (to_vec #(n - 1) (a / 2)) (Seq.create 1 (a % 2 = 1)) i;
+    assert (index #bool #n (to_vec a) i == index #bool #(n - 1) (to_vec #(n - 1) (a / 2)) i);
+    mod_pow2_div2 a m;
+    to_vec_mod_pow2 #(n - 1) (a / 2) (m - 1) i
+    end
+
+val to_vec_lt_pow2: #n:size_nat -> a:U.uint_t n -> m:nat -> i:nat{i < n - m} ->
+  Lemma (requires (a < pow2 m))
+        (ensures  (index #bool #n (to_vec a) i == false))
+        [SMTPat (index #bool #n (to_vec #n a) i); SMTPat (pow2 m)]
+let rec to_vec_lt_pow2 #n a m i =
+  if n = 0 then ()
+  else
+    if m = 0 then
+      assert (a == U.zero n)
+    else
+      begin
+      lemma_index_app1 (to_vec #(n - 1) (a / 2)) (Seq.create 1 (a % 2 = 1)) i;
+      assert (index #bool #n (to_vec a) i == index #bool #(n - 1) (to_vec #(n - 1) (a / 2)) i);
+      to_vec_lt_pow2 #(n - 1) (a / 2) (m - 1) i
+      end
+
+val logxor_disjoint: #n:size_nat -> a:U.uint_t n -> b:U.uint_t n -> m:pos{m < n} ->
+  Lemma (requires (a % pow2 m == 0 /\ b < pow2 m))
+        (ensures  (U.logxor #n a b == a + b))
+
+#push-options "--z3rlimit 20 --max_fuel 0"
+let logxor_disjoint #n a b m =
+  assert (forall (i:nat{n - m <= i /\ i < n}).{:pattern (index #bool #n (to_vec a) i)}
```

Comment:
are you familiar with `calc` statements? they might make this sort of proof a lot more readable

Reply:
This function is ported from `logor_disjoint` in `FStar.UInt.fst` https://github.com/FStarLang/FStar/blob/master/ulib/FStar.UInt.fst#L302
The body of the function was hand-copied from there, I still can improve the syntax for better readability.